### DISCUSSION TITLE
I can separate Admin app from API server?

### DISCUSSION THREAD

**Discussion by rahadurr** [directus/directus#6739] [Archive]
I would like to know that is there any way to separate Admin UI from the API server. I want to set up two different URL for my API and admin web such as:

https://api.example.com
https://admin.example.com

**Comment by rijkvanzanten** [comment]
No, the admin app is served by the API and can't be used standalone

**Comment by mpenna** [comment]
Sorry for resurrecting such an old thread, but the impossibility of such a separation, as proposed by the OP, is rather intriguing me. I gather that the D8 app is a full-blown Vue/Vite application which proxies its requests to the D8 api, isn't it? So why would this separation not be possible? If my assumption is correct, I think this separation could enable a few other hosting scenarios.

**Reply by rijkvanzanten** [reply]
The separation isn't possible as the API injects HTML into the app when it's served, and because of third-party cookie / CORS policies that are common in browsers now (configuration of those are non-trivial nowadays, and caused a lot of headaches / questions in v8 before) 👍🏻

**Reply by mpenna** [reply]
Thanks for the quick reply. OK, I got it. Perhaps this multi-tier architecture could eventually be revisited in the near future. It would be nice to have the admin app be a separate tier, maybe a simple statically-built SPA that could be served by a CDN or something along those lines.

**Reply by xu4wang** [reply]
> The separation isn't possible as the API injects HTML into the app when it's served, and because of third-party cookie / CORS policies that are common in browsers now (configuration of those are non-trivial nowadays, and caused a lot of headaches / questions in v8 before) 👍🏻

I did some test before, running the APP in dev mode, by changing the vite.config.js using another API server instance.

```js
	server: {
		port: 8080,
		proxy: {
			'^/(?!admin)': {
				target: process.env.API_URL ? process.env.API_URL : 'http://localhost:8055/',
				changeOrigin: true,
			},
		},
		fs: {
			allow: [searchForWorkspaceRoot(process.cwd()), ...getExtensionsRealPaths()],
		},
	},
```

I was even able to use an API server in the cloud.  The APP in dev mode seems running without any issue.

Is this a configuration the OP was asking for?

**Reply by domdus** [reply]
To have two different domains, I would do with NGINX VHosts only. But I needed an external api with long cache TTL and an internal one with no cache. (different config)

I run some sort of separated setup by launching two instances of directus. (side effects, see bottom)
One instance is the public api only! - no admin.
The other serves the admin app - In my case only accessible internally in a private VLAN via VPN. Technically this runs the admin + api again for itself.

You can create two directus config files e.g. config_app.js and config_api.js (.js / .json / .yaml).
(https://docs.directus.io/self-hosted/config-options.html#configuration-files)
Both files have almost same config (DB connection, secrets, storage, ...), but for the api instance you configure the following:

- Different port (eg.8060)
- ROOT_REDIRECT: false
- SERVE_APP: false
- Different cache TTL?

Now you launch two directus instances (out of a single directus installation) with the specific launch config file each:
`CONFIG_PATH="/path/to/config_app.js" npx directus start`
`CONFIG_PATH="/path/to/config_api.js" npx directus start`

(You can better/easy achieve this with pm2 in production)

Finally you need to create two VHosts (NGINX?) for both instances and its ports:
- https://api.example.com/ -> port 8060
- https://admin.example.com/ -> port 8055

---
BE AWARE OF POSSIBLE SIDE EFFECTS:
- Schema changes (deleted fields/collections) done in the app instance are not recognized by the other api instance (schema cache) and can cause heavy exceptions. You better restart the api app when you do such changes.
- When using flows cron jobs, a flow is triggered on both instances simultanousely

**Reply by xu4wang** [reply]
> 
> BE AWARE OF POSSIBLE SIDE EFFECTS:
> 
> * Schema changes (deleted fields/collections) done in the app instance are not recognized by the other api instance (schema cache) and can cause heavy exceptions. You better restart the api app when you do such changes.


@domdus thanks for the warnning.. ;-)

would you please detail on teh "schema cache" issue?  where is the "schema cache" stored? and how it is being kept?

I run into some problem when I delete/recreate a database table from the API endpoints extensions using low level knex instance. 
I am suspecting it is related to "schema cache" you mentioned.  Since the table I re-created is with exactly the same structure, but I need to restart directus sometimes to make the app recognise the table again.

**Reply by domdus** [reply]
I have no detailed experience with the schema cache itself, but meanwhile there are at least a lot more options to deal with it.
Did not find time yet to check things, but take a look here:

https://docs.directus.io/self-hosted/config-options.html#cache
`CACHE_SYSTEM_TTL`
`CACHE_SCHEMA`
The schema caching behavior can be configured seperately now.

Also there is a quite new utils api endpoint to clear this cache programmatically:
https://docs.directus.io/reference/system/utilities.html#clear-the-internal-cache
Maybe it can be used in a flow when schema changes happen.

Regarding the "were" and "how" I only know that there are some changes made in the most recent directus update:
https://github.com/directus/directus/pull/17763 `Rework schema caching to use memory`

**Reply by luciandavidescu** [reply]
@mpenna see my answer below

**Comment by mpenna** [comment]
@xu4wang Would you care to detail a bit more how you got everything working? I tried something along those lines yesterday, but couldn't get it to work.

**Reply by xu4wang** [reply]
@mpenna I just did some test, the purpose of the test was to check how to modify the source of the app adding more customization. It was not in production.

First,  I cloned the directus repo locally, and change the vite.config.js under app folder to:

```js
	server: {
		port: 8080,
		proxy: {
			'^/(?!admin)': {
				target: process.env.API_URL ? process.env.API_URL : 'https://direcuts.at.clould/',
				changeOrigin: true,
			},
		},
		fs: {
			allow: [searchForWorkspaceRoot(process.cwd()), ...getExtensionsRealPaths()],
		},
	},
```

https://direcuts.at.clould:8080/ is another server running directus in the cloud.

After that, I run `pnpm --filter @directus/app dev`  and I can access all the features of the app from my local vite dev env, with the APIs connected to the clould.

**Reply by mpenna** [reply]
Thanks @xu4wang I did try something along those lines but had some issues getting both tiers working together. Probably some misconfiguration related to my local Dockerized setup. I'll try to revisit this later on to see if I can get it working as intended. Let me clarify a few extra details, though: 

- which instance is running the API?
- is the API configured to serve the app or did you disable that setting (`SERVE_APP=false`)?

**Reply by xu4wang** [reply]
> 
> - which instance is running the API?
> 

the one in the cloud.

> - is the API configured to serve the app or did you disable that setting (`SERVE_APP=false`)?

No. it is serving its own APP too.

**Comment by luciandavidescu** [comment]
One workaround that may suit most purposes is to duplicate the API via the webserver (have two routes point to the same location) to get;
1) a 'private' route – https://admin.example.com – which consumes itself for its own admin things;
2) a 'public' route – https://api.example.com – for the things that need to be exposed, but with the admin path blocked.

_With nginx, add these blocks to the public server's location to simulate the api response:_

	location /admin {
	error_page 404 @message;
	}

	location @message {
	default_type application/json;
	set $new_uri "";if ($request_uri ~ "^([^?]*)(\?.*)?$") {set $new_uri $1;}
	return 404 '{"errors":[{"message":"Route $new_uri doesn\'t exist.","extensions":{"code":"ROUTE_NOT_FOUND"}}]}';
	}

**Comment by Eggwise** [comment]
Any update on this ?
I want to build and package the admin frontend as a desktop application.
How do i build the frontend to static folder ?

@rijkvanzanten

**Reply by rijkvanzanten** [reply]
Directus isn't designed or built to have the application ran separately from the API. You could potentially fork, build it from source, and hack your way to success, but that's not officially supported nor recommended
